src/Symfony/Component/DependencyInjection/Compiler/FactoryReturnTypePass.php

```diff
@@ -0,0 +1,86 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection\Compiler;
+
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Definition;
+use Symfony\Component\DependencyInjection\Reference;
+
+/**
+ * @author Guilhem N. <egetick@gmail.com>
+ */
+class FactoryReturnTypePass implements CompilerPassInterface
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function process(ContainerBuilder $container)
+    {
+        // works only since php 7.0
+        if (!method_exists(\ReflectionMethod::class, 'getReturnType')) {
+            return;
+        }
+
+        foreach ($container->getDefinitions() as $id => $definition) {
+            $this->updateDefinition($container, $id, $definition);
+        }
+    }
+
+    private function updateDefinition(ContainerBuilder $container, $id, Definition $definition, array $previous = array())
+    {
+        // circular reference
+        if (isset($previous[$id])) {
+            return;
+        }
+
+        $factory = $definition->getFactory();
+        if (null === $factory || $definition->getClass()) {
+            return;
+        }
+
+        if (is_string($factory) && function_exists($factory)) {
+            $reflectionFunction = new \ReflectionFunction($factory);
+        } else {
+            if ($factory[0] instanceof Reference) {
+                $this->updateDefinition(
+                    $container,
+                    (string) $factory[0],
+                    $factoryDefinition = $container->findDefinition((string) $factory[0]),
+                    array_merge($previous, array($id => ''))
+                );
+                $class = $factoryDefinition->getClass();
+            } else {
+                $class = $factory[0];
+            }
+
+            try {
+                $reflectionFunction = new \ReflectionMethod($class, $factory[1]);
+            } catch (\ReflectionException $e) {
+                return;
+            }
+        }
+
+        $returnType = $reflectionFunction->getReturnType();
+        if (null !== $returnType && !$returnType->isBuiltin()) {
+            $returnType = (string) $returnType;
+            if (isset($class)) {
+                if ('self' === $returnType) {
+                    $returnType = $class;
```

Comment:
this is wrong. It needs to be resolved based on ``$reflectionFunction->getDeclaringClass()``, not based on the class being used. Otherwise it will be wrong for a child class